contracts/RefundProvider/RefundProvider.sol

```diff
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./RefundState.sol";
+import "../ProviderInterface/IProviderSingleIdRegistrar.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol";
+
+contract RefundProvider is RefundState, IERC721Receiver {
+    constructor(address nftContract, address provider) {
+        require(nftContract != address(0x0) && provider != address(0x0), "invalid address");
+        lockDealNFT = LockDealNFT(nftContract);
+        lockProvider = LockDealProvider(provider);
+    }
+
+    ///@dev refund implementation
+    function onERC721Received(
+        address provider,
+        address receiver,
+        uint256 poolId,
+        bytes calldata
+    ) external override returns (bytes4) {
+        require(msg.sender == address(lockDealNFT), "invalid nft contract");
+        if (provider == receiver) {
+            require(
+                lockProvider.startTimes(poolId - 1) > block.timestamp,
+                "too late"
+            );
+            DealProvider dealProvider = lockProvider.dealProvider();
+            lockDealNFT.transferFrom(address(this), lockDealNFT.ownerOf(poolId - 1), poolId - 2);
+            lockDealNFT.transferFrom(lockDealNFT.ownerOf(poolId - 1), receiver, poolId - 1);
+
+            lockDealNFT.setPoolIdToProvider(address(dealProvider), poolId - 2);
+            lockDealNFT.setPoolIdToProvider(address(dealProvider), poolId - 1);
+        }
+        return IERC721Receiver.onERC721Received.selector;
+    }
+
+    ///@param params[0] = tokenLeftAmount
+    ///@param params[params.length - 2] = refundMainCoinAmount
+    ///@param params[params.length - 1] = refund finish time
+    function createNewRefundPool(
+        address token,
+        address owner,
+        address mainCoin,
+        address provider,
+        uint256[] calldata params
+    ) external returns (uint256 poolId) {
+        uint256 paramsLength = params.length;
+        require(paramsLength > 2, "invalid params length");
+        require(lockDealNFT.isApprovedForAll(msg.sender, address(this)), "invalid approval");
```

Comment:
Do we need this check?

Reply:
I think so because the `RefundProvider` contract uses the project owner's nft (pool) during the refund process. If we take the case where the `RefundProvider` owns the pool, then we need to do two transfers or have access to the `vaultManager` or create more pools, which is not good for us.

Reply:
The project owner should be able to withdraw their main coins after the time limit using the `Withdraw` function

Reply:
That's strange logic to require the NFT allowance when creating a new pool. I think the current tests work fine without that check, so we can remove it for now and add it later on where it's needed (e,g, withdraw).

Reply:
if remove line 42 in the tests, they will fail
https://github.com/The-Poolz/LockDealNFT/blob/a711c7d4e73454f00017fab684d35df1031cf061/test/RefundProvider.ts#L42

Reply:
perhaps you are right in case the project owner at some point forbids the use of his pool. but in this case we need to change the `withdraw/refund` logic

Reply:
Okay, let's try to have the `RefundProvider` be the owner to deny the ability to close the refund option.

Reply:
if we want to make the refund the owner, we need make other provider for the project owner